pkg/services/ssosettings/api/api.go

```diff
@@ -91,18 +101,41 @@ func (api *Api) getAuthorizedList(ctx context.Context, identity identity.Request
 	return authorizedProviders, nil
 }
 
+// swagger:route GET /v1/sso-settings/{key} sso_settings getProviderSettings
+//
+// # Get an SSO Settings entry by Key
+//
+// You need to have a permission with action `settings:read` with scope `settings:auth.<provider>:*`.
+//
+// Responses:
+// 200: okResponse
+// 400: badRequestError
+// 401: unauthorisedError
+// 403: forbiddenError
+// 404: notFoundError
+// 500: internalServerError
 func (api *Api) getProviderSettings(c *contextmodel.ReqContext) response.Response {
 	key, ok := web.Params(c.Req)[":key"]
 	if !ok {
 		return response.Error(http.StatusBadRequest, "Missing key", nil)
 	}
 
-	settings, err := api.SSOSettingsService.GetForProvider(c.Req.Context(), key)
+	provider, err := api.SSOSettingsService.GetForProvider(c.Req.Context(), key)
+
 	if err != nil {
-		return response.Error(http.StatusNotFound, "The provider was not found", err)
+		if errors.Is(err, ssosettings.ErrNotFound) {
+			return response.Error(http.StatusNotFound, "The provider was not found", err)
+		}
+		return response.Error(http.StatusInternalServerError, "Failed to get provider settings", err)
 	}
 
-	return response.JSON(http.StatusOK, settings)
+	if clientSecret, ok := provider.Settings["ClientSecret"].(string); ok && len(clientSecret) > 0 {
+		provider.Settings["ClientSecret"] = "*********"
+	}
```

Comment:
This should be a generic implementation for any protocol (oauth, saml, ldap) and provider. So I don't think it'a a good idea to refer settings by their name here (`ClientSecret`). There is a function in the sso service to identify the  fields that should be hidden: https://github.com/grafana/grafana/blob/main/pkg/services/ssosettings/ssosettingsimpl/service.go#L238. Maybe you can use that?

Reply:
Thanks, I forgot to go back and make this more generic instead of hardcoding for OAuth.

Reply:
Make sure you also remove this section since you added the section above that covers the same thing.